README.markdown
jQuery-typing with events and data-api support
=============

Assign callbacks for started/stopped typing events. This version supports events and data-api and is backwards compatible with original version by narf.


Usage (events automatically attached just-in-time based on the data API - bootstrap style)
-----

  <input data-provide="typing" data-typing-delay="400" />
  $(body).bind('typing:start', ':input', function(event) { $(this).css('background', '#fa0'); });
  $(body).bind('typing:start', ':input', function(event) { $(this).css('background', '#f00'); });

Alternative Usage 
-----

    $(':text').typing({
        start: function (event, $elem) {
            $elem.css('background', '#fa0');
        },
        stop: function (event, $elem) {
            $elem.css('background', '#f00');
        },
        delay: 400
    });

Add `data-provide="typing"` attribute to your input and it will auto-magically start triggering `typing:start` and `typing:stop` events - you can treat these as ony other DOM events. This comes handy when working with frameworks like [Backbone JS][].

`typing` command takes key-value object with `start`, `stop` and
`delay` keys. They are all optional, so you can either pass only
`start` callback, `stop` callback, `stop` callback and `delay` time,
or everything.

`delay` is amount of time the plugin waits for another keypress before
judging that typing has stopped; it is expressed in milliseconds and
defaults to 400. Regardless of `delay`'s value, the `stop` callback is
called immediately when blur event occurs.

Callbacks are passed two arguments: event that caused callback execution
and jQuery object for matched element. Possible events are `keypress`
or `keydown` for `start` callbacks and `keyup` or `blur` for `stop`
callbacks.


Demo
----

Visit <http://narf.pl/jquery-typing/>


Download
--------

Get production version from & development version visit [GitHub][].

  [GitHub]: http://github.com/tnajdek/jquery-typing


Meta
----

jQuery-typing is written by [Maciej Konieczny][] and uses
[semantic versioning][] for release numbering.  Everything in `plugin/`
directory is released into the [public domain][].

jQuery-typing has been tweaked by [Tom Najdek][] to support data-api and trigger events.

  [Maciej Konieczny]: http://narf.pl/
  [Tom Najdek]: http://doppnet.com/
  [semantic versioning]: http://semver.org/
  [public domain]: http://unlicense.org/
  [Backbone JS]: http://backbonejs.org/


HISTORY.markdown
jQuery-typing history
=====================


This file only covers changes important for the end-user.  Visit GitHub
for [full history][].

  [full history]: http://github.com/tnajdek/jquery-typing/commits/master

[v0.3.1][]
--------
  
  - Fixed bug where event listener would be attached multiple times

  - Settings can be changed dynamically via api after typing has been attached to a DOM element

[v0.3][]
--------

  - Added support for data-api.

  - Added events triggering

[v0.2][]
--------

  - Added callback arguments support.  They are passed two things: event
    that caused callback execution and jQuery object for matched element.

  - Added blur support.


[v0.1][]
--------

  - Initial release.

  [v0.3.1]: http://github.com/tnajdek/jquery-typing/compare/v0.3.0...v0.3.1
  [v0.3]: http://github.com/tnajdek/jquery-typing/compare/v0.2.0...v0.3.0
  [v0.2]: http://github.com/narfdotpl/jquery-typing/compare/v0.1.0...v0.2.0
  [v0.1]: http://github.com/narfdotpl/jquery-typing/compare/c4b8c10...v0.1.0


plugin/jquery.typing-0.3.0.js
// jQuery-typing
//
// Version: 0.3.0
// Website: http://narf.pl/jquery-typing/
// License: public domain <http://unlicense.org/>
// Author:  Maciej Konieczny <hello@narf.pl>
// Events & data-api: Tom Najdek <tom@doppnet.com>

(function ($) {

	//--------------------
	//  jQuery extension
	//--------------------

	$.fn.typing = function (options) {
		return this.each(function (i, elem) {
			listenToTyping(elem, options);
		});
	};


	//-------------------
	//  actual function
	//-------------------

	function listenToTyping(elem, options) {
		// override default settings
		var settings = $.extend({
			start: null,
			stop: null,
			delay: 400
		}, options);

		// create other function-scope variables
		var $elem = $(elem),
			typing = false,
			delayedCallback;

		// start typing
		function startTyping(event) {
			if (!typing) {
				// set flag and run callback
				typing = true;
				$elem.trigger('typing:start');
				if (settings.start) {
					settings.start(event, $elem);
				}
			}
		}

		// stop typing
		function stopTyping(event, delay) {
			if (typing) {
				// discard previous delayed callback and create new one
				clearTimeout(delayedCallback);
				delayedCallback = setTimeout(function () {
					// set flag and run callback
					typing = false;
					$elem.trigger('typing:stop');
					if (settings.stop) {
						settings.stop(event, $elem);
					}
				}, delay >= 0 ? delay : settings.delay);
			}
		}

		// listen to regular keypresses
		$elem.keypress(startTyping);

		// listen to backspace and delete presses
		$elem.keydown(function (event) {
			if (event.keyCode === 8 || event.keyCode === 46) {
				startTyping(event);
			}
		});

		// listen to keyups
		$elem.keyup(stopTyping);

		// listen to blurs
		$elem.blur(function (event) {
			stopTyping(event, 0);
		});
	}

	//provide data-api bootstrap style (http://rc.getbootstrap.com/javascript.html)
	$(document).on('focus.typing.data-api', '[data-provide=typing]', function (e) {
		var $this = $(this),
		delay = $this.data('typingDelay');
		$this.typing( {
			delay: delay
		});
	});

})(jQuery);


plugin/jquery.typing-0.3.1.js
// jQuery-typing
//
// Version: 0.3.0
// Website: http://narf.pl/jquery-typing/
// License: public domain <http://unlicense.org/>
// Author:  Maciej Konieczny <hello@narf.pl>
// Events & data-api: Tom Najdek <tom@doppnet.com>

(function ($) {

	//--------------------
	//  jQuery extension
	//--------------------

	$.fn.typing = function (options) {
		return this.each(function (i, elem) {
			var	$elem = $(elem),
				api;
			
			if(!$elem.data('typing')) {
				api = new Typing(elem, options);
				$elem.data('typing', api);
			}
		});
	};


	//-------------------
	//  actual function
	//-------------------

	var Typing = function(elem, options) {
		// create other function-scope variables
		var $elem = $(elem),
			typing = false,
			delayedCallback,
		// override default settings
			settings = $.extend({
				start: null,
				stop: null,
				delay: 400
			}, options);

		//export settings to the api
		this.settings = settings;
	

		// start typing
		function startTyping(event) {
			if (!typing) {
				// set flag and run callback
				typing = true;
				$elem.trigger('typing:start');
				if (settings.start) {
					settings.start(event, $elem);
				}
			}
		}

		// stop typing
		function stopTyping(event, delay) {
			if (typing) {
				// discard previous delayed callback and create new one
				clearTimeout(delayedCallback);
				delayedCallback = setTimeout(function () {
					// set flag and run callback
					typing = false;
					$elem.trigger('typing:stop');
					if (settings.stop) {
						settings.stop(event, $elem);
					}
				}, delay >= 0 ? delay : settings.delay);
			}
		}

		// listen to regular keypresses
		$elem.keypress(startTyping);

		// listen to backspace and delete presses
		$elem.keydown(function (event) {
			if (event.keyCode === 8 || event.keyCode === 46) {
				startTyping(event);
			}
		});

		// listen to keyups
		$elem.keyup(stopTyping);

		// listen to blurs
		$elem.blur(function (event) {
			stopTyping(event, 0);
		});
	};

	//provide data-api bootstrap style (http://rc.getbootstrap.com/javascript.html)
	$(document).on('focus.typing.data-api', '[data-provide=typing]', function (e) {
		var $this = $(this),
		delay = $this.data('typingDelay');
		$this.typing( {
			delay: delay
		});
	});

})(jQuery);
